subsys/bluetooth/mesh/provisioner.c

```diff
@@ -0,0 +1,767 @@
+#include "adv.h"
+#include "mesh.h"
+#include "net.h"
+#include "rpl.h"
+#include "beacon.h"
+#include "access.h"
+#include "foundation.h"
+#include "proxy.h"
+#include "prov.h"
+#include "settings.h"
+
+static struct {
+	struct bt_mesh_cdb_node *node;
+	uint16_t addr;
+	uint16_t net_idx;
+	uint8_t attention_duration;
+	uint8_t uuid[16];
+} prov_device;
+
+static void send_pub_key(void);
+static void prov_dh_key_gen(void);
+static void pub_key_ready(const uint8_t *pkey);
+
+static int reset_state(void)
+{
+	if (prov_device.node != NULL) {
+		bt_mesh_cdb_node_del(prov_device.node, false);
+	}
+
+	return bt_mesh_prov_reset_state(pub_key_ready);
+}
+
+static void prov_link_close(enum prov_bearer_link_status status)
+{
+	BT_DBG("%u", status);
+	bt_mesh_prov_link.expect = PROV_NO_PDU;
+
+	bt_mesh_prov_link.bearer->link_close(status);
+}
+
+static void prov_fail(uint8_t reason)
+{
+	/* According to Bluetooth Mesh Specification v1.0.1, Section 5.4.4, the
+	 * provisioner just closes the link when something fails, while the
+	 * provisionee sends the fail message, and waits for the provisioner to
+	 * close the link.
+	 */
+	prov_link_close(PROV_BEARER_LINK_STATUS_FAIL);
+}
+
+static void send_invite(void)
+{
+	PROV_BUF(inv, 2);
+
+	BT_DBG("");
+
+	bt_mesh_prov_buf_init(&inv, PROV_INVITE);
+	net_buf_simple_add_u8(&inv, prov_device.attention_duration);
+
+	bt_mesh_prov_link.conf_inputs[0] = prov_device.attention_duration;
+
+	if (bt_mesh_prov_send(&inv, NULL)) {
+		BT_ERR("Failed to send invite");
+		return;
+	}
+
+	bt_mesh_prov_link.expect = PROV_CAPABILITIES;
+}
+
+static void start_sent(int err, void *cb_data)
+{
+	if (!bt_pub_key_get()) {
+		atomic_set_bit(bt_mesh_prov_link.flags, WAIT_PUB_KEY);
+		BT_WARN("Waiting for local public key");
+	} else {
+		send_pub_key();
+	}
+}
+
+static void send_start(uint8_t method, uint8_t action, uint8_t size)
+{
+	BT_DBG("");
+
+	PROV_BUF(start, 6);
+
+	const uint8_t *data = &bt_mesh_prov_link.conf_inputs[1 + 3];
+
+	bt_mesh_prov_buf_init(&start, PROV_START);
+	net_buf_simple_add_u8(&start, PROV_ALG_P256);
+
+	if (atomic_test_bit(bt_mesh_prov_link.flags, REMOTE_PUB_KEY) &&
+	    *data == PUB_KEY_OOB) {
+		net_buf_simple_add_u8(&start, PUB_KEY_OOB);
+		atomic_set_bit(bt_mesh_prov_link.flags, OOB_PUB_KEY);
+	} else {
+		net_buf_simple_add_u8(&start, PUB_KEY_NO_OOB);
+	}
+
+	net_buf_simple_add_u8(&start, bt_mesh_prov_link.oob_method);
+
+	net_buf_simple_add_u8(&start, bt_mesh_prov_link.oob_action);
+
+	net_buf_simple_add_u8(&start, bt_mesh_prov_link.oob_size);
+
+	memcpy(&bt_mesh_prov_link.conf_inputs[12], &start.data[1], 5);
+
+	if (bt_mesh_prov_auth(method, action, size) < 0) {
+		BT_ERR("Invalid authentication method: 0x%02x; "
+		       "action: 0x%02x; size: 0x%02x", method,
+		       action, size);
+		return;
+	}
+
+	if (bt_mesh_prov_send(&start, start_sent)) {
+		BT_ERR("Failed to send Provisioning Start");
+		return;
+	}
+}
+
+static void prov_check_method(struct bt_mesh_prov_capabilities *caps)
```

Comment:
This method is called "check_method", but it's also sending start. The code would be easier to follow if this function simply returned true/false, and `prov_capabilities` checked this before calling `send_start`/`prov_failed`. `send_start` can use `bt_mesh_prov_link.oob_method` and similar parameters as input.

Reply:
I thought about it, but I didn't.